File: api/src/utils/jwt.rs
use chrono::{Duration, Utc};
use jsonwebtoken::{decode, encode, DecodingKey, EncodingKey, Header, TokenData, Validation};
use serde::{Deserialize, Serialize};
use axum::http::StatusCode;

use crate::utils::constants;

#[derive(Serialize, Deserialize)]
pub struct Claims {
    pub exp: u64,
    pub iat: u64,
    pub is_client: bool,
    pub email: String,
}

pub fn encode_jwt(email: String, is_client: bool) -> Result<String, StatusCode> {
    let now = Utc::now();
    let expire = Duration::days(30);

    let claims = Claims { 
        exp: (now+expire).timestamp() as u64, 
        iat: now.timestamp() as u64, 
        email, 
        is_client,
    };
    let secret = (*constants::TOKEN).clone();

    encode(
        &Header::default(), 
        &claims, 
        &EncodingKey::from_secret(secret.as_ref())
    )
    .map_err(|_| StatusCode::INTERNAL_SERVER_ERROR)
}

pub fn decode_jwt(jwt: String) -> Result<TokenData<Claims>, StatusCode> {
    let secret = (*constants::TOKEN).clone();

    decode(
        &jwt, 
        &DecodingKey::from_secret(secret.as_ref()), 
        &Validation::default()
    )
    .map_err(|_| StatusCode::INTERNAL_SERVER_ERROR)
}

File: backend/src/util/jwt.ts
import jwt from 'jsonwebtoken';
import { TOKEN } from './constants';

export interface Claims {
    exp: number;
    iat: number;
    is_client: boolean;
    email: string;
}

export function encodeJWT(email: string, is_client: boolean): string {
    const now = Math.floor(Date.now() / 1000);
    const expire = 30 * 24 * 60 * 60;

    const claims: Claims = {
        exp: now + expire,
        iat: now,
        email,
        is_client,
    };

    if (!TOKEN) throw new Error('TOKEN não definido');

    return jwt.sign(claims, TOKEN);
}

export function decodeJWT(token: string): Claims | null {
    if (!TOKEN) throw new Error('TOKEN não definido');

    try {
        const decoded = jwt.verify(token, TOKEN) as Claims;
        return decoded;
    } catch (err) {
        return null;
    }
}
